sdk-setup/src/mb2

```diff
@@ -1791,20 +1791,103 @@ run_build_requires__process_args() {
         BUILD_REQUIRES_MODE=pull
     fi
 }
+
 sign_rpms() {
     [[ -e $STATEDIR/spec ]] || fatal "No previous build found"
     local rpms=$(find_rpms $all_opt "$OPT_OUTPUTDIR" "$STATEDIR/spec")
     [[ $rpms ]] || fatal "No package found"
+
     local sdksign=/usr/libexec/sdk-setup/sdk-rpmsig
+
     if [[ -x $sdksign ]]; then
         $sdksign ${rpms}
     else
-        rpmsign --addsign ${rpms}
+        sign_rpms_native ${rpms}
     fi
+
     local rc=$?
     [[ $rc -eq 0 ]] || { notice "Failed to sign packages"; return $rc; }
 }
 
+sign_rpms_native() {
+    local rpms=("$@")
+
+    local tmpdir=
+    sign_rpms_native_cleanup() (
+        trap 'echo cleaning up...' INT TERM HUP
+        if [[ $tmpdir ]]; then
+            rm -rf "$tmpdir"
+        fi
+    )
+    trap 'sign_rpms_native_cleanup; trap - RETURN' RETURN
+    trap 'return 1' INT TERM HUP
+
+    local extra_args=()
+
+    if [[ $OPT_PACKAGE_SIGN_USER ]]; then
+
+        if inside_build_engine; then
+            local shared_key=${OPT_SHARED_DIR}/gnupg/${OPT_PACKAGE_SIGN_USER}.key
+            if [[ ! -f $shared_key ]]; then
+                notice "Cannot sign packages: Internal error: No GPG key shared for the user '$OPT_PACKAGE_SIGN_USER'."
+                return 1
+            fi
+
+            shared_key_id=$(gpg2 --with-colons --list-packets "$shared_key" | grep -oP ':user ID packet: (.*)' | cut -d "\"" -f2)
+
+            keys_fingerprint=($(gpg2 --list-secret-keys --fingerprint "$shared_key_id" | grep -Po "Key fingerprint = \K(.+)" | sed 's/ //g'))
+            for fingerprint in $keys_fingerprint; do
+                if ! gpg2 --batch --delete-secret-keys "$fingerprint"; then
+                    notice "Cannot sign packages: Failed to delete existing '$fingerprint' key."
+                    return 1
+                fi
+            done
+
+            if ! gpg2 --import "${shared_key}" &> /dev/null; then 
+                notice "Cannot sign packages: Failed to import key from file '$shared_key'."
+                return 1
+            fi
```

Comment:
- local variables
- error handling
- gpg2 use without --with-colons - output format not stable
- wrong dereference of array variable
- public keys not deleted
- use gpg2 --import with --quiet instead of suppressing all output, including possibly valuable error output

```suggestion
            local shared_key_id=$(gpg2 --with-colons --list-packets "$shared_key" \
                    |sed -n 's/:user ID packet: "\(.*\)"/\1/p')
            [[ $shared_key_id ]] || return

            local key_fingerprints=$(gpg2 --with-colons --list-secret-keys --fingerprint "$shared_key_id" \
                    |awk -F: '$1 == "fpr" { print $10 }')
            local key_fingerprint=
            for key_fingerprint in $key_fingerprints; do
                if ! gpg2 --batch --delete-secret-and-public-key "$key_fingerprint"; then
                    notice "Cannot sign packages: Internal error: Failed to delete old GPG key copy '$key_fingerprint'."
                    return 1
                fi
            done

            if ! gpg2 --quiet --import "${shared_key}"; then 
                notice "Cannot sign packages: Internal error: Failed to import GPG key from file '$shared_key'."
                return 1
            fi
```